structure/redblacktree/rbtree.go

```diff
@@ -0,0 +1,404 @@
+// rbtree.go
+//
+// description: A struct representing a red black tree
+//
+// details:
+//
+// A Red Black tree is a self-balancing Binary Search Tree.
+// It considers an additional attribute for each node, a color either red or black
+// By requiring that no two consecutive nodes in a Depth Traversal can have red color,
+// the tree ensures that the heights of all subtrees remain close enough
+// to ensure O(n) time search
+//
+// This file contains the tree with a single field - root node
+//
+// There are multiple methods, public ones are
+// Insert - inserts a value
+// Search - searches and returns Node pointer
+// Delete - deletes a value
+// PrintInorder - prints inorder representation - Val and isRed
+// PrintPreorder - prints preorder representation - Val and isRed
+//
+// Implementation is based on Introduction to Algorithms, Cormen et. al., 3rd edition
+//
+// Wikipedia: https://en.wikipedia.org/wiki/Red%E2%80%93black_tree
+// Geeks-for-geeks series: https://www.geeksforgeeks.org/red-black-tree-set-1-introduction-2/
+//
+// author: VinWare (https://github.com/VinWare)
+
+package redblacktree
+
+// For printing
+import (
+	"fmt"
+)
+
+type RBTree struct {
+	Root *Node
+}
+
+func (t *RBTree) Insert(Val int) {
+	t.insert(Val)
+	t.Root.isRed = false
+}
+
+// Internal insert function, refer Intro to Algorithms mentioned above
+func (t *RBTree) insert(Val int) {
```

Comment:
Could you break this function down into separable pieces? Currently its responsibility is to insert a node but its finding the empty node to act as parent (which can be taken out of this function as a function itself) and doing so many things that its not readable (and indeed understandable) at all. Taking care of different cases can easily be handled by a strategy pattern instead of creating complicated if and else statements. Additionally, it seems that you have just re-written code from this book you mentioned and translated to Go. Keep in mind that some of the concepts in other languages make it difficult to write algorithms in Go and if you don't adapt the algorithm itself it will be very messy. Right now, the algorithm is so convoluted that it took me more than 50mins to properly understand this - which shouldn't be the case because Red Black tree is essentially a Binary search tree and they are supposed to be easy to understand.

Reply:
Insert is now broken down into - finding a parent (`findNewParent`, the main logic (setting Root if needed else setting parent's children node values), and the tree rebalancing part (`rebalanceTree`). I have rewritten `rotate` to avoid duplicating code between left and right rotates, and because of this, I can remove the separate`if-else` statements for left and right. Now there are just one level of `if-else` statements in the `climbTreeColoring` method which I feel are simpler than a strategy pattern